event/consumer.go

```diff
@@ -2,79 +2,123 @@ package event
+func (consumer *Consumer) Consume(
+	ctx context.Context,
+	messageConsumer MessageConsumer,
+	handler Handler,
+	cfg *config.Config) {
 
-	// consume loop, to be executed by each worker
-	var consume = func(workerID int) {
+	go func() {
+		defer close(consumer.closed)
+
+		batch := NewBatch(cfg.BatchSize)
+		// Wait a batch full of messages.
+		// If we do not get any messages for a time, just process the messages already in the batch.
 		for {
 			select {
-			case message, ok := <-messageConsumer.Channels().Upstream:
-				if !ok {
-					log.Info(ctx, "closing event consumer loop because upstream channel is closed", log.Data{"worker_id": workerID})
-					return
+			case msg := <-messageConsumer.Channels().Upstream:
+				AddMessageToBatch(ctx, batch, msg, handler)
+				msg.Release()
+
+			case <-time.After(cfg.BatchWaitTime):
+				if batch.IsEmpty() {
+					continue
 				}
-				messageCtx := context.Background()
-				processMessage(messageCtx, message, handler, cfg)
-				message.Release()
-			case <-messageConsumer.Channels().Closer:
-				log.Info(ctx, "closing event consumer loop because closer channel is closed", log.Data{"worker_id": workerID})
+				log.Info(ctx, "batch wait time reached. proceeding with batch", log.INFO,
+					log.Data{
+						"batch-wait-time": cfg.BatchWaitTime,
+						"batchsize":       batch.Size(),
+					})
+				ProcessBatch(ctx, handler, batch, "timeout")
+
+			case eventClose := <-consumer.closing:
+				log.Info(eventClose.ctx, "closing event consumer loop")
+				close(consumer.closing)
 				return
 			}
 		}
-	}
+	}()
+}
 
-	// workers to consume messages in parallel
-	for w := 1; w <= cfg.KafkaNumWorkers; w++ {
-		go consume(w)
+// AddMessageToBatch will attempt to add the message to the batch and determine if it should be processed.
+func AddMessageToBatch(ctx context.Context, batch *Batch, msg kafka.Message, handler Handler) {
+	log.Info(ctx, "add message to batch starts")
+	batch.Add(ctx, msg)
+	if batch.IsFull() {
+		ProcessBatch(ctx, handler, batch, "full-batch")
 	}
 }
 
-// processMessage unmarshals the provided kafka message into an event and calls the handler.
-// After the message is handled, it is committed.
-func processMessage(ctx context.Context, message kafka.Message, handler Handler, cfg *config.Config) {
-
-	// unmarshal - commit on failure (consuming the message again would result in the same error)
-	event, err := unmarshal(message)
+// ProcessBatch will attempt to handle and commit the batch, or shutdown if something goes horribly wrong.
+func ProcessBatch(ctx context.Context, handler Handler, batch *Batch, reason string) {
+	log.Info(ctx, "process batch starts", log.Data{
+		"process batch starts time": time.Now(),
+		"batch_size":                batch.Size(),
+		"reason":                    reason})
+	err := handler.Handle(ctx, batch.Events())
 	if err != nil {
-		log.Error(ctx, "failed to unmarshal event", err)
-		message.Commit()
+		log.Error(ctx, "error processing batch", err)
```

Comment:
Should we not be calling `batch.Commit()` here, otherwise the batch will continue to increase in size?

Also I will create a new ticket for how we handle failure scenario here - we will want to requeue all events/messages in batch due to failure.